### ISSUE TITLE
Towards a list of modules for the new wallet

### ISSUE THREAD

**Issue by rbrunner7** [seraphis-migration/wallet3#27]
@DangerousFreedom1984 wrote in issue #23:

------------

1\) Design choice. Currently wallet2 and simplewallet are the big two beasts when we talk about the Monero wallet. There are a few classes but almost no modularity. I don't think it is performing bad but it is definitely hard to read and understand what is going on. I believe we could do a bit better and introduce some modularity by separating the methods into some groups like in the figure below. Basically I would like to have classes (or only namespaces maybe) that groups these methods for organizational purposes and keep only the program flow and all the important private variables at Wallet3.
![class diagram](https://user-images.githubusercontent.com/98208928/198821968-22b16fe7-819e-4401-a512-226f037e9167.png)

**Comment by rbrunner7** [comment]
@UkoeHB commented:

-------------------

`wallet2`'s problem is not just lack of modularity, it is an over-burdened API. We should really have three or more 'wallet' objects to separate normal, multisig, and view-only interfaces.

**Comment by rbrunner7** [comment]
@DangerousFreedom1984 reaction to this:

------------------

That's a very useful comment. Thanks. Yeah right now we just have one struct in the account class to handle everything. I totally agree it would be better to split the cases. I'm not fully aware of the implications though, does it also mean that ideally we would not have any conversion between normal/multisig/view-only wallets like it happens today?

**Comment by rbrunner7** [comment]
I learned from the "Seraphis code walkthrough" video that koe has made a list of interfaces defined in his library that we need to implement, in his issue #8, chapter "Mock-ups to fully implement".

That's not yet a list of modules for our new wallet, but contains some pointers in that direction.

**Comment by rbrunner7** [comment]
Maybe that's something we can discuss in the meeting tomorrow, but when I am faced with a complex new task like this new wallet, and I am not yet able to come up with a proper list of modules, at some point I stop to worry and start to implement in a "bottom-up" style: I take something that I **can** identify as a needed module for sure, and implement that. Usually doing so has the effect that I learn more about the job as a whole, so it's no problem to continue with more and more modules in that way, until building the whole is not much more difficult than putting Lego bricks together - ideally :)

I suspect that the Seraphis library also came to be in this style at least in part. Correct, @UkoeHB ?

**Comment by UkoeHB** [comment]
>implement in a "bottom-up" style: I take something that I can identify as a needed module for sure, and implement that. 

Yes this is more-or-less the approach I took when building the seraphis library. And then refactoring various pieces as I learned more about what they need to do and how to structure them better.
